### PR TITLE
Paying to each holder in each group

### PR DESCRIPTION
**Pull request by joshuef**
starting to simplify payments to a per chunk basis

A first step in #638## Description

### Summary generated by Reviewpad on 16 Aug 23 14:29 UTC
This pull request includes the following changes:
1. In the `lib.rs` file:
   - Import `PublicKey` from `libp2p::identity` and use it in addition to `Multiaddr` from `libp2p`.
   - Add a new field `reward_main_key` of type `PublicKey` to the struct `Node`.
2. In the `file_apis.rs` file:
   - Import statement added: `use crate::NetworkAddress`.
   - Import statement removed: `use crate::messages::PaymentProof`.
   - Import statement added: `use sn_dbc::DbcId`.
   - Method added: `network_address()`, returns the `NetworkAddress` associated with the chunk.
   - Struct field type changed: `payment` changed from `PaymentProof` to `Vec<DbcId>`.
3. In the `api.rs` file:
   - Added `PublicKey` import from `identity` module.
   - Renamed method `get_store_cost_from_network` to `get_store_costs_from_network` and changed its return type to `Result<Vec<(PublicKey, Token)>>`.
   - Updated error handling and added decoding of `PublicKey` in `get_store_costs_from_network` method.
   - Added a retry loop for `put_record_with_retries` method.
   - Made `get_closest_peers` method public.
4. In the `storage_payments.rs` file:
   - Import `sn_dbc::{Hash, Token}` has been removed.
   - The type of `random_content_addrs` has been changed from `XorName` to `sn_protocol::NetworkAddress::ChunkAddress`.
   - The `eyre::Result` import has been simplified.
   - Various changes related to the variable `invalid_proofs` have been made in the function `storage_payment_chunk_upload_fails`.
5. In the `mod.rs` file in the `sn_protocol/src/messages` directory:
   - In the `use` statement, `PaymentProof` has been replaced with `PaymentTransactions`.
6. In the `data_with_churn.rs` file:
   - The method `pay_for_storage` is modified to use the `network_address()` method instead of the `name()` method on each element of the `chunks` iterator.
7. In the `verify_data_location.rs` file:
   - There is a change in the `store_chunk` function where the `pay_for_storage` method is called with the `network_address()` method instead of the `name()` method on each element of the `chunks` iterator.
8. In the `sn_cli/src/subcommands/files.rs` file:
   - The import statement `sn_transfers::wallet::PaymentProofsMap` has been replaced with `sn_transfers::wallet::PaymentTransactionsMap`.
   - The argument `payment_proofs` in the function `upload_chunks` has been changed from `&PaymentProofsMap` to `&PaymentTransactionsMap`.
9. In the `wallet.rs` file:
   - The type `PaymentProofsMap` has been renamed to `PaymentTransactionsMap`.
   - The `get_payment_proof` method now accepts a `NetworkAddress` parameter instead of `XorName`, and returns an `Option<&Vec<DbcId>>` instead of `Option<&PaymentProof>`.
   - The `add_payment_proofs` method now accepts a `PaymentTransactionsMap` parameter instead of `PaymentProofsMap`.
   - The `local_send_storage_payment` method now accepts a `storage_cost_for_addr` parameter of type `Token` instead of `storage_payment` and `root_hash`.
   - The `wallet.paymet_proofs` field has been renamed to `wallet.payment_transactions`.
10. In the `.github/workflows/memcheck.yml` file:
    - Line 101: The `timeout-minutes` value has been updated from 25 to 40.
11. In the `sn_networking/src/error.rs` file:
    - Added `use libp2p::identity::DecodingError;`.
    - Added variant `Libp2pDecode` to the `Error` enum with a `DecodingError` parameter.
12. In the `sn_protocol/src/messages/cmd.rs` file:
    - The struct `PaymentProof` has been renamed to `PaymentTransactions`.
    - The fields `audit_trail` and `path` of the `PaymentTransactions` struct have been removed.
13. In the `sn_client` module:
    - The `network_store_cost` field has been removed from the `Client` struct.
14. In the `transfer.rs` file:
    - Line 66: The `root_hash` variable is no longer passed as a parameter to the function `create_storage_payment_transfer()`.
    - Line 70: Added a comment to indicate that some code needs to be cleaned up in the future.
    - Line 72: The `root_hash` variable is now created using the `sn_dbc::Hash::hash()` function and the value "nonsense" as input.
15. In the `sn_protocol/src/messages/mod.rs` file:
    - The import statement for `sn_protocol::storage::ChunkAddress` has been removed.

Please review these changes and let me know if you have any questions or need further assistance.

### CODE REVIEW

File: sn_client/src/api.rs [hunk 1]
```diff
@@ -315,27 +313,28 @@ impl Client {
...
             value: try_serialize_record(&[spend.signed_spend], RecordKind::DbcSpend)?,
             publisher: None,
             expires: None,
         };
-        Ok(self.network.put_record(record, verify_store).await?)
+        Ok(self.network.put_record(record, verify_store, true).await?)
     }
 
     /// Get a dbc spend from network
+    /// Optionally pass in an expected Dbc if we're verifying a spend from the client
```
Comment (RolandSherwin) [inline sn_client/src/api.rs]: Do we still optionally pass in a Dbc?

File: sn_client/src/wallet.rs [hunk 2]
```diff
@@ -104,86 +103,103 @@ impl WalletClient {
...
-        if addrs_to_pay.is_empty() {
-            trace!("We already hold payment proofs for all the records.");
-            return Ok((proofs, None));
+    ) -> Result<(TransferOutputsMap, Token)> {
+        let mut total_cost = Token::zero();
+
+        let mut payment_map = BTreeMap::default();
+
+        let mut tasks = JoinSet::new();
+        // we can collate all the payments together into one transfer
```
Comment (RolandSherwin) [inline sn_client/src/wallet.rs:119]: What do we mean by one transfer here? We do 1 transfer per record right?

File: sn_client/src/wallet.rs [hunk 3]
```diff
@@ -104,86 +103,103 @@ impl WalletClient {
...
-            self.unconfirmed_txs.push(transfer);
-            return Err(error);
+        trace!("Sending storage payment transfer to the network");
+
+        for transfer in transfer_map.values() {
+            // first push all to the network, then try again and verify.
+            // otherwise all pushes are waiting on all prior verifications
+            self.client.send(transfer.clone(), false).await?;
+
         }
```
Comment (RolandSherwin) [inline sn_client/src/wallet.rs]: Maybe for the follow up PR? I think we can parallelize the `sends` here?

File: sn_networking/src/error.rs [hunk 4]
```diff
@@ -64,6 +68,8 @@ pub enum Error {
 
     #[error("Dial Error")]
     DialError(#[from] DialError),
+    #[error("Libp2p Idendity Decode Error")]
```
Comment (RolandSherwin) [inline sn_networking/src/error.rs]:
```suggestion
    #[error("Libp2p Identity Decode Error")]
```

File: sn_networking/src/lib.rs [hunk 5]
```diff
@@ -704,6 +708,13 @@ impl Network {
                             && target_record == Some(returned_record.clone()))
                     {
                         return Ok(returned_record);
+                    } else if expect_no_prior_record
+                        && (target_record.is_some()
+                            && target_record != Some(returned_record.clone()))
+                    {
+                        error!("Record check: expected {target_record:?}");
+                        error!("Returned: {returned_record:?}");
+                        return Err(Error::ExpectedNoRecordToExist(returned_record.key.into()));
```
Comment (RolandSherwin) [inline sn_networking/src/lib.rs]: Should we also return error if `target_record.is_none()`? 
Also can we add some docs for this argument?

File: sn_protocol/src/error.rs [hunk 6]
```diff
@@ -67,11 +67,11 @@ pub enum Error {
...
     /// Payment proof received has no inputs
-    #[error("Payment proof received for {0:?} has no inputs in its transaction")]
-    PaymentProofWithoutInputs(XorName),
+    #[error("Payment proof received for {0:?} has no dbc for this node in its transaction")]
+    NoPaymentToThisNode(XorName),
     /// The id of the fee output found in a storage payment proof is invalid
-    #[error("The id of the fee output found in a storage payment proof is invalid: {}", .0.to_hex())]
-    PaymentProofInvalidFeeOutput(Hash),
+    #[error("The id of the fee output found in a storage payment proof is invalid: {0:?}")]
+    PaymentProofInvalidFee(Token),
```
Comment (RolandSherwin) [inline sn_protocol/src/error.rs]: This error is unused and can be removed.

File: sn_client/src/wallet.rs (hunk 2 above)
Comment (joshuef) [reply sn_client/src/wallet.rs:119]: yeh, one per record is what we have right now. That could be improved upon I think

File: sn_client/src/wallet.rs (hunk 3 above)
Comment (joshuef) [reply sn_client/src/wallet.rs]: hmm, things get complicated here, as they may actually be chained. I thin I'm going to remove this and we have to look at what makes sense to do in parallel here as optimisations down the line.

File: sn_node/src/put_validation.rs [hunk 7]
```diff
@@ -230,16 +231,32 @@ impl Node {
...
         let validated_spends = match self
-            .signed_spend_validation(signed_spends, dbc_id, present_locally)
+            .signed_spend_validation(signed_spends.clone(), dbc_id, present_locally)
             .await?
         {
             Some(spends) => spends,
             None => {
                 // data is already present
-                return Ok(CmdOk::DataAlreadyPresent);
+                if is_new_put {
```
Comment (joshuef) [inline sn_node/src/put_validation.rs:240]: if it's a new put, can the data already be present? I'm not sure I follow this change

Comment (maqi) [reply sn_node/src/put_validation.rs:240]: If it's a new put, it's just follow the current flow.
The put could be fired by mistake or say `second put try to ensure data uploaded`

**Pull request closed without merging**